provider/azure/instance_information.go

```diff
@@ -0,0 +1,33 @@
+package azure
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/instances"
+)
+
+var _ environs.InstanceTypesFetcher = (*azureEnviron)(nil)
+
+// InstanceTypes implements InstanceTypesFetcher
+func (env *azureEnviron) InstanceTypes(c constraints.Value) (instances.InstanceTypesWithCostMetadata, error) {
+	types, err := env.getInstanceTypes()
+	if err != nil {
+		return instances.InstanceTypesWithCostMetadata{}, errors.Trace(err)
+	}
+	result := make([]instances.InstanceType, len(types))
+	i := 0
+	for _, iType := range types {
+		result[i] = iType
+		i++
+	}
+	result, err = instances.MatchingInstanceTypes(result, "", c)
```

Comment:
I wasn't suggesting you move this into each provider. You can/should do this outside of each provider, but you *also* need the constraints in the provider-level API because *some* providers will need to generate results based on those constraints. i.e. the constraints are required for generic filtering, but also for generating custom instance types.